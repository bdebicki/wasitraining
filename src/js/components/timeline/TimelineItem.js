import DATA_URL from '../../enums/data';
import TIMELINE from './elementHandlers/timeline';
import { addSVGmask, svgType } from '../../utils/addSvgMask';
import Edition from '../../classes/Edition';
import YearView from '../../views/YearView';

export default class TimelineItem extends Edition {
  static getEditionIdFromHref(link) {
    return parseInt(link.getAttribute('href').replace('#edition', ''), 10);
  }

  switchView(e) {
    const id = TimelineItem.getEditionIdFromHref(this);

    e.preventDefault();

    fetch(DATA_URL)
      .then((response) => response.json())
      .then((data) => {
        const year = new YearView(data, id);

        year.switchToYearView();
      })
      .catch((error) => {
        console.log(error); // eslint-disable-line no-console
      });
  }

  switchEdition(e) {
    const id = TimelineItem.getEditionIdFromHref(this);

    e.preventDefault();

    fetch(DATA_URL)
      .then((response) => response.json())
      .then((data) => {
        const year = new YearView(data, id);

        year.updateDetails(this);
      })
      .catch((error) => {
        console.log(error); // eslint-disable-line no-console
      });
  }

  renderYearMask() {
    const { editionYear } = this;
    const textMask = document.createElementNS(svgType, 'text');

    textMask.classList.add(TIMELINE.MAIN_EDITION_MASK_TEXT_CLASS);
    textMask.setAttributeNS(null, 'x', '50%');
    textMask.setAttributeNS(null, 'y', '50%');
    textMask.textContent = editionYear;

    const maskOptions = {
      svgClass: TIMELINE.MAIN_EDITION_MASK_CLASS,
      maskId: `yearMask${editionYear}`,
      maskShape: textMask,
      maskBgClass: TIMELINE.MAIN_EDITION_MASK_BG_CLASS,
    };

    return addSVGmask(maskOptions);
  }

  renderYear() {
    const span = document.createElement('span');

    span.classList.add(TIMELINE.MAIN_EDITION_YEAR_CLASS);
    span.textContent = this.editionYear;

    return span;
  }

  renderMainLink() {
    const a = document.createElement('a');
    const year = this.renderYear();
    const mask = this.renderYearMask();

    a.href = `#edition${this.editionId}`;
    a.classList.add(TIMELINE.MAIN_EDITION_LINK_CLASS);
    a.addEventListener('click', this.switchView, null);
    a.appendChild(year);
    a.appendChild(mask);

    return a;
  }

  renderNavLink(isActive) {
    const a = document.createElement('a');

    a.href = `#edition${this.editionId}`;
    a.classList.add(TIMELINE.NAV_EDITION_LINK_CLASS);
    if (isActive) {
      a.classList.add(TIMELINE.NAV_EDITION_ACTIVE_CLASS);
    }
    a.addEventListener('click', this.switchEdition, null);
    a.textContent = this.editionYear;

    return a;
  }

  renderMainEdition() {
    const li = document.createElement('li');

    li.classList.add(TIMELINE.MAIN_EDITION_CLASS);
    li.appendChild(this.renderMainLink());

    return li;
  }

  renderNavEdition(isActive) {
    const li = document.createElement('li');

    li.classList.add(TIMELINE.NAV_EDITION_CLASS);
    li.appendChild(this.renderNavLink(isActive));

    return li;
  }
}
