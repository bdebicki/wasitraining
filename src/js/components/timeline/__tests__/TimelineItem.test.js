import editionData from '../../../../../tests/__mocks__/edition-alphabeticalExceptHeadliners-noMergeArtists.json';
import TimelineItem from '../TimelineItem';

const timelineItem = new TimelineItem(editionData);

describe('timeline item component', () => {
  it('render main timeline item', () => {
    expect(timelineItem.renderMainEdition()).toMatchSnapshot();
  });
  it('render nav timeline item', () => {
    expect(timelineItem.renderNavEdition()).toMatchSnapshot();
  });
  it('render active nav timeline item', () => {
    expect(timelineItem.renderNavEdition(true)).toMatchSnapshot();
  });
});
