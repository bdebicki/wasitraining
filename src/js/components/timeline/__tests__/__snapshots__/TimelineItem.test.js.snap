// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeline item component render active nav timeline item 1`] = `
<li
	class="navTimelineItem"
>
	<a
		class="navTimelineItem__link navTimelineItem__link--isActive"
		href="#edition16"
	>
		2017
	</a>
</li>
`;

exports[`timeline item component render main timeline item 1`] = `
<li
	class="mainTimelineItem"
>
	<a
		class="mainTimelineItem__link"
		href="#edition16"
	>
		<span
			class="mainTimelineItem__year"
		>
			2017
		</span>
		<svg
			class="svgMask mainTimelineItemMask"
		>
			<defs>
				<mask
					id="yearMask2017"
					maskUnits="userSpaceOnUse"
				>
					<rect
						class="svgMask__base"
					/>
					<text
						class="mainTimelineItemMask__text"
						x="50%"
						y="50%"
					>
						2017
					</text>
				</mask>
			</defs>
			<rect
				class="svgMask__bg mainTimelineItemMask__bg"
				mask="url(#yearMask2017)"
			/>
		</svg>
	</a>
</li>
`;

exports[`timeline item component render nav timeline item 1`] = `
<li
	class="navTimelineItem"
>
	<a
		class="navTimelineItem__link"
		href="#edition16"
	>
		2017
	</a>
</li>
`;
